fenix/app/src/main/java/org/mozilla/fenix/debugsettings/store/DebugDrawerNavigationMiddleware.kt

```diff
@@ -0,0 +1,40 @@
+/* This Source Code Form is subject to the terms of the Mozilla Public
+ * License, v. 2.0. If a copy of the MPL was not distributed with this
+ * file, You can obtain one at http://mozilla.org/MPL/2.0/. */
+
+package org.mozilla.fenix.debugsettings.store
+
+import androidx.navigation.NavController
+import mozilla.components.lib.state.Middleware
+import mozilla.components.lib.state.MiddlewareContext
+import org.mozilla.fenix.debugsettings.navigation.DEBUG_DRAWER_ROUTE_LANDING_SCREEN
+import org.mozilla.fenix.debugsettings.navigation.DEBUG_DRAWER_ROUTE_TAB_TOOLS
+
+/**
+ * Middleware that handles navigation events for the Debug Drawer feature.
+ *
+ * @param navController [NavController] used to execute any navigation actions on the UI.
+ */
+class DebugDrawerNavigationMiddleware(
+    private val navController: NavController,
+) : Middleware<DebugDrawerState, DebugDrawerAction> {
+
+    override fun invoke(
+        context: MiddlewareContext<DebugDrawerState, DebugDrawerAction>,
+        next: (DebugDrawerAction) -> Unit,
+        action: DebugDrawerAction,
+    ) {
+        when (action) {
+            is DebugDrawerAction.NavigateTo.LandingScreen -> {
+                navController.navigate(route = DEBUG_DRAWER_ROUTE_LANDING_SCREEN)
```

Comment:
Please excuse my complete lack of understanding here. I am wondering why we wanto to do navigation via middleware compare to the Compose Navigation examples of defining a `NavHost` and just navigating directly.

Reply:
Kinda similar reasons why we would have, historically, done the navigation from within Interactors/Controllers.
- Keep the navigation calls outside of the UI (the nav controller will still be connected to a `NavHost` in follow-up work noted above)
- Have something that's testable

Reply:
Makes sense. My initial impression is that it seems unnecessarily complicated.
Some followup questions:

1. I am wondering if we really have all that much test coverage around our Compose code for the lambda/callback functions. I understand the idea is to ideally just dispatch actions in these callback and handle this all in our middlewares. It makes me wonder if navigation is really something we care about having test coverage (if we did have good test coverage for our composables).
2. Do we also have examples of this being done elsewhere already?

Reply:
hope that all makes sense. this is definitely the most technically & architecturally challenging part of this feature.

Reply:
Thanks for taking the time for these responses, I do really appreciate. I am just not sure if I am still suffering from smooth brain or just being really far detached from the code since I have been doing for monorepo for so long.

Reply:
sure thing [ReviewQualityCheckNavigationMiddleware](https://searchfox.org/mozilla-mobile/source/firefox-android/fenix/app/src/main/java/org/mozilla/fenix/shopping/middleware/ReviewQualityCheckNavigationMiddleware.kt)

It uses a `UseCase` in this example.

I'm going to do a deep dive on all of this soon(tm), but I'm always happy to do a call sometime to catch up on the current state/future state with ya.